File: runtimes/dotnet-6.0/src/RuntimeContext.cs
using System.Collections; 
using System.Text.Json;

namespace DotNetRuntime
{
	public class RuntimeContext
	{
		public RuntimeRequest Req { get; set; }
		public RuntimeResponse Res { get; set; }

		public List<string> Logs { get; private set; } = new List<string>();
		public List<string> Errors { get; private set; } = new List<string>();

		public RuntimeContext(RuntimeRequest req, RuntimeResponse res)
		{
			Req = req;
			Res = res;
		}

		public void Log(object message)
		{
			if (message is IList || message is IDictionary) {
				Logs.Add(JsonSerializer.Serialize(message));
			} else {
				Logs.Add(message.ToString() ?? "");
			}
		}

		public void Error(object message)
		{
			if (message is IList || message is IDictionary) {
				Errors.Add(JsonSerializer.Serialize(message));
			} else {
				Errors.Add(message.ToString() ?? "");
			}
		}
	}
}

File: runtimes/java-11.0/src/main/java/io/openruntimes/java/RuntimeContext.java
package io.openruntimes.java;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class RuntimeContext {
    private final RuntimeRequest req;
    private final RuntimeResponse res;

    private final List<String> logs = new ArrayList<>();
    private final List<String> errors = new ArrayList<>();

    public RuntimeRequest getReq() {
        return this.req;
    }

    public RuntimeResponse getRes() {
        return this.res;
    }

    public List<String> getLogs() {
        return this.logs;
    }

    public List<String> getErrors() {
        return this.errors;
    }

    private static final Gson gson = new GsonBuilder().serializeNulls().create();

    public RuntimeContext(
            RuntimeRequest req,
            RuntimeResponse res
    ) {
        this.req = req;
        this.res = res;
    }

    public void log(Object message) {
        if (message instanceof Map || message instanceof List || message instanceof Set) {
            this.logs.add(gson.toJson(message));
        } else {
            this.logs.add(message.toString());
        }
    }

    public void error(Object message) {
        if (message instanceof Map || message instanceof List || message instanceof Set) {
            this.errors.add(gson.toJson(message));
        } else {
            this.errors.add(message.toString());
        }
    }
}


